src/Repositories/RedirectRepository.php

```diff
@@ -0,0 +1,390 @@
+use Bonnier\WP\Redirect\Database\Migrations\Migrate;
+use Bonnier\WP\Redirect\Database\Query;
+use Bonnier\WP\Redirect\Exceptions\IdenticalFromToException;
+use Bonnier\WP\Redirect\Helpers\LocaleHelper;
+use Bonnier\WP\Redirect\Helpers\UrlHelper;
+use Bonnier\WP\Redirect\Models\Redirect;
+use Illuminate\Support\Collection;
+
+class RedirectRepository extends BaseRepository
+{
+    /**
+     * RedirectRepository constructor.
+     * @param DB $database
+     * @throws \Exception
+     */
+    public function __construct(DB $database)
+    {
+        $this->tableName = Migrate::REDIRECTS_TABLE;
+        parent::__construct($database);
+    }
+
+    /**
+     * @param int $redirectID
+     * @return Redirect|null
+     */
+    public function getRedirectById(int $redirectID): ?Redirect
+    {
+        $data = $this->database->findById($redirectID);
+        if ($data) {
+            $redirect = new Redirect();
+            return $redirect->fromArray($data);
+        }
+
+        return null;
+    }
+
+    /**
+     * @param string $path
+     * @param string|null $locale
+     * @return Redirect|null
+     * @throws \Exception
+     */
+    public function findRedirectByPath(string $path, string $locale = null): ?Redirect
+    {
+        if ($redirect = $this->findExactRedirectByPath($path, $locale)) {
+            return $this->handleQueryParams($redirect, $path);
+        }
+
+        if ($redirect = $this->findRedirectByIgnoringQueryParams($path, $locale)) {
+            return $this->handleQueryParams($redirect, $path);
+        }
+
+        if ($redirect = $this->findWildcardRedirectsByPath($path, $locale)) {
+            return $this->handleQueryParams($redirect, $path);
+        }
+
+        return null;
+    }
+
+    /**
+     * @param string $path
+     * @param string|null $locale
+     * @return Redirect|null
+     * @throws \Exception
+     */
+    public function findExactRedirectByPath(string $path, string $locale = null): ?Redirect
+    {
+        if (is_null($locale)) {
+            $locale = LocaleHelper::getLanguage();
+        }
+
+        $normalizedPath = UrlHelper::normalizePath($path);
+
+        $query = $this->database->query()->select('*')
+            ->where(['from_hash', hash('md5', $normalizedPath)])
+            ->andWhere(['locale', $locale]);
+        $results = $this->database->getResults($query);
+
+        if ($results) {
+            $redirect = new Redirect();
+            return $redirect->fromArray($results[0]);
+        }
+
+        return null;
+    }
+
+    /**
+     * @param string $path
+     * @param string|null $locale
+     * @return Redirect|null
+     * @throws \Exception
+     */
+    public function findRedirectByIgnoringQueryParams(string $path, string $locale = null): ?Redirect
+    {
+        if (is_null($locale)) {
+            $locale = LocaleHelper::getLanguage();
+        }
+
+        if (parse_url($path, PHP_URL_QUERY)) {
```

Comment:
Is this check necessary? surely you do not care if the path has query or not?

Reply:
Yes, the check is necessary. It's a recursive function through `findRedirectByPath()`
Adding a comment, to clear up what's happening.